imap/monitor.go

```diff
@@ -0,0 +1,194 @@
+				break
+			}
+			if len(imapSettings) > 0 {
+				im := imapSettings[0]
+				// 3. Check if IMAP is enabled
+				if im.Enabled {
+					log.Debug("Checking IMAP for user ", uid, ": ", im.Username, "@", im.Host)
+					checkForNewEmails(im)
+					time.Sleep((time.Duration(im.IMAPFreq) - 10) * time.Second) // Subtract 10 to compensate for the default sleep of 10 at the bottom
+				}
+			}
+		}
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// NewMonitor returns a new instance of imap.Monitor
+func NewMonitor() *Monitor {
+
+	im := &Monitor{}
+	return im
+}
+
+// Start launches the IMAP campaign monitor
+func (im *Monitor) Start() error {
+	log.Info("Starting IMAP monitor manager")
+	ctx, cancel := context.WithCancel(context.Background()) // ctx is the derivedContext
+	im.cancel = cancel
+	go im.start(ctx)
+	return nil
+}
+
+// Shutdown attempts to gracefully shutdown the IMAP monitor.
+func (im *Monitor) Shutdown() error {
+	log.Info("Shutting down IMAP monitor manager")
+	im.cancel()
+	return nil
+}
+
+// checkForNewEmails logs into an IMAP account and checks unread emails
+//  for the rid campaign identifier.
+func checkForNewEmails(im models.IMAP) {
+
+	im.Host = im.Host + ":" + strconv.Itoa(int(im.Port)) // Append port
+	mailSettings := MailboxInfo{
+		Host:   im.Host,
+		TLS:    im.TLS,
+		User:   im.Username,
+		Pwd:    im.Password,
+		Folder: im.Folder}
+
+	msgs, err := GetUnread(mailSettings, true, false)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	// Update last_succesful_login here via im.Host
+	err = models.SuccessfulLogin(&im)
+
+	if len(msgs) > 0 {
+		var reportingFailed []uint32 // UIDs of emails that were unable to be reported to phishing server, mark as unread
+		var campaignEmails []uint32  // UIDs of campaign emails. If DeleteReportedCampaignEmail is true, we will delete these
+		for _, m := range msgs {
+			// Check if sender is from company's domain, if enabled. TODO: Make this an IMAP filter
+			if im.RestrictDomain != "" { // e.g domainResitct = widgets.com
+				splitEmail := strings.Split(m.From.Address, "@")
+				senderDomain := splitEmail[len(splitEmail)-1]
+				if senderDomain != im.RestrictDomain {
+					log.Debug("Ignoring email as not from company domain: ", senderDomain)
+					continue
+				}
+			}
+
+			body := string(append(m.Text, m.HTML...)) // Empty body is being returned for emails forwarded from ProtonMail. Need to investigate more. Perhaps it's being attached, and eazye doesn't support attachments.
+			rid := goPhishRegex.FindString(body)
+
+			if rid != "" {
+				rid = rid[5:]
+				log.Infof("User '%s' reported email with rid %s", m.From.Address, rid)
+				result, err := models.GetResult(rid)
+				if err != nil {
+					log.Error("Error reporting GoPhish email with rid ", rid, ": ", err.Error())
+					reportingFailed = append(reportingFailed, m.UID)
+				} else {
+					err = result.HandleEmailReport(models.EventDetails{}) // Not sure if we should populate something into the EventDetails{}
```

Comment:
> `// Not sure if we should populate something into the EventDetails{}`

Nah, since there isn't any payload sending an empty `EventDetails` is correct.